pkg/datasource/accessors.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datasource
+
+import (
+	"errors"
+	"fmt"
+	"slices"
+)
+
+// FieldAccessor grants access to the underlying buffer of a field
+type FieldAccessor interface {
+	Name() string
+	Size() uint32
+	Get(Data) []byte
+	Set(Data, []byte) error
+	IsRequested() bool
+	AddSubField(name string, opts ...FieldOption) (FieldAccessor, error)
+	GetSubFieldsWithTag(tag ...string) []FieldAccessor
+	Parent() FieldAccessor
+	SubFields() []FieldAccessor
+	SetHidden(bool)
+}
+
+type fieldAccessor struct {
+	ds *dataSource
+	f  *field
+}
+
+func (a *fieldAccessor) Name() string {
+	return a.f.Name
+}
+
+func (a *fieldAccessor) Size() uint32 {
+	return a.f.Size
+}
+
+func (a *fieldAccessor) Get(d Data) []byte {
+	if FieldFlagEmpty.In(a.f.Flags) {
+		return nil
+	}
+	if a.f.Size > 0 {
+		// size must be valid here
+		return d.(*data).Payload[a.f.PayloadIndex][a.f.Offs : a.f.Offs+a.f.Size]
+	}
+	return d.(*data).Payload[a.f.PayloadIndex]
+}
+
+func (a *fieldAccessor) SetHidden(val bool) {
+	a.ds.lock.Lock()
+	defer a.ds.lock.Unlock()
+	if !val {
+		FieldFlagHidden.RemoveFrom(&a.f.Flags)
+	} else {
+		FieldFlagHidden.AddTo(&a.f.Flags)
+	}
+}
+
+func (a *fieldAccessor) Set(d Data, b []byte) error {
+	if FieldFlagEmpty.In(a.f.Flags) {
+		return errors.New("field cannot contain a value")
+	}
+	if a.f.Size != 0 && uint32(len(b)) != a.f.Size {
+		return fmt.Errorf("invalid size, expected %d, got %d", a.f.Size, len(b))
+	}
+	d.(*data).Payload[a.f.PayloadIndex] = b
```

Comment:
Why is it so different from the `Get()` implementation?

I was trying to implement a toy operator that changes the value of the pid field:

```go
func (c *myOpInstance) Start(gadgetCtx operators.GadgetContext) error {
	for _, ds := range gadgetCtx.GetDataSources() {
		fmt.Printf("subscribing to %s\n", ds.Name())

		pidAcc := ds.GetField("pid")
		ds.Subscribe(func(ds datasource.DataSource, data datasource.Data) error {
			pidBytes := make([]byte, 4)
			binary.LittleEndian.PutUint32(pidBytes, 31415926)
			pidAcc.Set(data, pidBytes)

			//pid := binary.LittleEndian.Uint32(pidAcc.Get(data))
			//fmt.Printf("pid: %d\n", pid)

			return nil
		}, 500)

	}

	return nil
}
```

But it crashes with 


```bash 
subscribing to events
panic: runtime error: slice bounds out of range [:313] with capacity 4

goroutine 1909 [running]:
github.com/inspektor-gadget/inspektor-gadget/pkg/datasource.(*fieldAccessor).Get(0xc004eb5f60, {0x4b57130?, 0xc004e48ea0?})
	/home/mvb/kinvolk/ebpf/inspektor-gadget/pkg/datasource/accessors.go:56 +0x108

```

It seems to me the current `Set()` method is only used to set the full data coming from ebpf (which contains multiple fields added with `AddStaticFields`), however I think the accessor returned by `ds.GetField()` should work fine when only modifying that field.

I think the following fixes it:

```go
func (a *fieldAccessor) Set(d Data, b []byte) error {
	if FieldFlagEmpty.In(a.f.Flags) {
		return errors.New("field cannot contain a value")
	}
	if a.f.Size != 0 && uint32(len(b)) != a.f.Size {
		return fmt.Errorf("invalid size, expected %d, got %d", a.f.Size, len(b))
	}
	// This is trying to set the full payload
	if a.f.Offs == 0 {
		d.(*data).Payload[a.f.PayloadIndex] = b
		return nil
	}
	copy(d.Raw().Payload[a.f.PayloadIndex][a.f.Offs:a.f.Offs+a.f.Size], b)
	return nil
}
```

But perhaps it's better to have two implementations of "fieldAccesor", (1) for single fields (2) for multiple fields returned by `AddStaticFields`.

Reply:
You're right, the branch handling that case is missing; I think the proper solution would be to add a flag like `StaticMember` that determines how to access it - `Offset` alone is not enough, because the first member would also have that set to zero.

Reply:
Added this now.